proposals/extensions_v2.md

```diff
@@ -0,0 +1,356 @@
+
+New extension members (beyond instance members) will need to have their metadata form decided on.  Consumption from older compilers and different languages of these new members will be specified at a later point in time.
+
+A full example of this translation with a real world complex signature would be:
+
+```c#
+static class Enumerable
+{
+    public static TResult Sum<TSource, TResult, TAccumulator>(this IEnumerable<TSource> source, Func<TSource, TResult> selector)
+        where TResult : struct, INumber<TResult>
+        where TAccumulator : struct, INumber<TAccumulator>
+    {
+        // Original body
+    }
+}
+
+extension Enumerable
+{
+    public TResult Sum<TSource, TResult, TAccumulator>(Func<TSource, TResult> selector)
+        for IEnumerable<TSource> source
+        where TResult : struct, INumber<TResult>
+        where TAccumulator : struct, INumber<TAccumulator>
+    {
+        // Exactly the same code as original body.
+    }
+}
+```
+
+This form supports *non* extension static methods.  For example `Enumerable.Range` would migrate like so:
+
+```c#
+static class Enumerable
+{
+    public static IEnumerable<int> Range(int start, int count) { ... }
+}
+
+
+extension Enumerable
+{
+    // Exact same signature.  No 'for-clause'.
+    public static IEnumerable<int> Range(int start, int count) { ... }
+}
+```
+
+## Disambiguation
+
+Classic extension methods today can be disambiguated by falling back to static-invocation syntax.  For example, if `x.Count()` is ambiguous, it is possible to switch to some form of `StaticClass.Count(x)` to call the desired method.  A similar facility is needed for modern extension methods.  While the existing method-invocation-translation approach works fine for methods (where the receiver can be remapped to the first argument of the static extension method call), it is ungainly for these other extension forms.
+
+As an initial strawman this proposal suggests reusing `cast expression` syntax for disambiguation purposes.  For example:
+
+```c#
+var v1 = ((Extension)receiver).ExtensionMethod(); // instead of Extension.ExtensionMethod(receiver)
+var v2 = ((Extension)receiver).ExtensionProperty;
+var v3 = ((Extension)receiver)[indexerArg];
+var v4 = (Extension)receiver1 + receiver2;
+```
+
+Constructors and static methods would not need any special syntax as the extension can cleanly be referenced as a type where needed.
+
+```c#
+var v3 = new Extension(...); // Makes instance of the actual extended type.
+var v4 = Extension.StaticExtensionMethod(...);
+```
+
+Note 1: while the cast syntax traditionally casts or converts a value, that would not be the case for its use here.  It would only be used as a lookup mechanism to indicate which extension gets priority.  Importantly, even with this syntax, extensions themselves are not types.  For example:
```

Comment:
Is `typeof(Extension)` also blocked? `nameof(Extension)`?

Reply:
great points.  no.  neither woudl be blocked.  will call out.